File: python/server/routes/cache.py
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from ..database import get_db
from ..models.cache import CacheEntry
from ..utils import decode_token

router = APIRouter(prefix="/api/cache")


async def authenticate(request: Request):
    auth = request.headers.get("authorization", "")
    payload = decode_token(auth)
    if not payload:
        raise HTTPException(status_code=401, detail="Unauthorized")
    return int(payload["sub"])


@router.get("/colors/{album_id}")
async def get_colors(album_id: str, request: Request, session: AsyncSession = Depends(get_db)):
    await authenticate(request)
    result = await session.execute(
        select(CacheEntry).where(CacheEntry.cache_key == f"album_colors:{album_id}")
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Not found")
    return {"colors": entry.value}


@router.get("/album/{album_id}")
async def get_album(album_id: str, request: Request, session: AsyncSession = Depends(get_db)):
    await authenticate(request)
    result = await session.execute(
        select(CacheEntry).where(CacheEntry.cache_key == f"album_feather:{album_id}")
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Not found")
    return {"image": entry.value}


@router.get("/name/{song_uri}")
async def get_song_name(song_uri: str, request: Request, session: AsyncSession = Depends(get_db)):
    await authenticate(request)
    result = await session.execute(
        select(CacheEntry).where(CacheEntry.cache_key == f"song_name:{song_uri}")
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Not found")
    return {"name": entry.value}


@router.post("/precache")
async def precache(request: Request, session: AsyncSession = Depends(get_db)):
    await authenticate(request)
    return {"success": True}

File: js/apps/server/src/routes/cache.ts
import { Router } from 'express';
import { prisma } from '../utils/database.js';
import { asyncHandler } from '../utils/http.js';
import { getUserIdFromAuthHeader } from '../services/authService.js';

export const cacheRouter = Router();

cacheRouter.use(
  asyncHandler(async (req, _res, next) => {
    await getUserIdFromAuthHeader(req.headers.authorization);
    next();
  })
);

async function getEntryValue(key: string): Promise<string | null> {
  const entry = await prisma.cacheEntry.findUnique({ where: { key } });
  return entry?.value ?? null;
}

function parseJsonSafe(raw: string | null): unknown {
  if (raw === null) return null;
  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

cacheRouter.get(
  '/colors/:albumId',
  asyncHandler(async (req, res) => {
    const value = await getEntryValue(`album_colors:${req.params.albumId}`);
    if (value === null) {
      res.status(404).json({ detail: 'Not found' });
      return;
    }
    res.json({ colors: parseJsonSafe(value) });
  })
);

cacheRouter.get(
  '/album/:albumId',
  asyncHandler(async (req, res) => {
    const value = await getEntryValue(`album_feather:${req.params.albumId}`);
    if (value === null) {
      res.status(404).json({ detail: 'Not found' });
      return;
    }
    res.json({ image: parseJsonSafe(value) });
  })
);

cacheRouter.get(
  '/name/:songUri',
  asyncHandler(async (req, res) => {
    const value = await getEntryValue(`song_name:${req.params.songUri}`);
    if (value === null) {
      res.status(404).json({ detail: 'Not found' });
      return;
    }
    const parsed = parseJsonSafe(value);
    res.json({ name: typeof parsed === 'string' ? parsed : JSON.stringify(parsed) });
  })
);

cacheRouter.post(
  '/precache',
  asyncHandler(async (_req, res) => {
    res.json({ success: true });
  })
);
